feat: add IDTokenCredential support

salrashid123 opened the pull request:
Fixes https://github.com/googleapis/google-auth-library-java/issues/99

Adds support for IDTokens into `google-auth-java` for `ServiceAccountCredentials`, `ComuteEngineCredentials` and `ImpersonatedCredentials` 

GCP users currently have to do these steps manually and outside of `google-auth-java`:   
   ref: https://github.com/salrashid123/salrashid123.github.io/tree/master/google_id_token#java
this change wraps the various flows into the library

I do not have any testcases here. I'm submitting this first for review of the api surface.  Usage for this change is shown here:
 https://github.com/googleapis/google-auth-library-java/issues/99#issuecomment-512899563

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,89 @@
+   *         audience.
+   */
+
+  IdToken idTokenWithAudience(String targetAudience, List<String> options);
```
Rather than a list of Strings, let's make the options into an enum

```suggestion
  IdToken idTokenWithAudience(String targetAudience, List<Option> options);
```

```java
enum Option {
    FORMAT_FULL("formatFull"),
    LICENSES_TRUE("licensesTrue"),
    INCLUDE_EMAIL("includeEmail");
}
```

salrashid123 replied:
done; good idea

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+  private GoogleCredentials sourceCredentials;
+  private final String transportFactoryClassName;
+  private String targetAudience;
+  private IdToken idToken;
```
We shouldn't need this field. If `IdToken` is an `AccessToken` we should re-use the access token.

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+  private static final String CLOUD_PLATFORM_SCOPE = "https://www.googleapis.com/auth/cloud-platform";
+
+  private GoogleCredentials sourceCredentials;
+  private final String transportFactoryClassName;
```
I'm not sure why this is necessary.

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+    return this.idToken;
+  }
+
+  public IdToken getIdToken() {
```
The user can call `getAccessToken()`

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,100 @@
+
+  private static final long serialVersionUID = -8514239465808977353L;
+
+  private final String tokenValue;
```
This is unnecessary (inherited).

salrashid123 replied:
done; used `super()`

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,100 @@
+   * @param jws            JsonWebSignature as object
+   * @param audience       List of the Audiences the idToken was issued for.
+   */
+  public IdToken(String tokenValue, JsonWebSignature jws) {
```
The constructor here shouldn't be public. We can surface it with `public static IdToken create(String)`. We can also parse the JsonWebSignature here rather than doing it everywhere else.
```suggestion
  private IdToken(String tokenValue) {
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,100 @@
+   *
+   * @return The raw access token string value.
+   */
+  public String getTokenValue() {
```
This is inherited so it's not necessary (we're setting the tokenValue via the parent constructor)

salrashid123 replied:
done, used `super.`

salrashid123 replied:
resolving comment; lmk if this isn't all correct

salrashid123 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public interface IdTokenProvider {
+
+  class IdTokenProviderException extends RuntimeException {
```
done.

salrashid123 commented:
ok, i've addressed some of the request and left a bunch unsolved with this commit as they are relaed to the bit mentioned here:

The biggest thing is if 

```java
public class IdToken extends AccessToken
```

is the best way or not

The `IdToken` a bit different than `AccessToken`...the former is a readable jwt token that customers can readily use as-is in an authorized http transport as shown below or with any third party http transport library by coaxing out the token `credential.getAccessToken()` and then attaching it manually to an `Authorization: Bearer` header.

However, customers will also want to evaluate and inspect claims embedded within the token (expiration, the aud: field, etc)
```java
  public JsonWebSignature getJsonWebSignature() {
    return jws;
  }
```
for example

```java
System.out.println(tokenCredential.getIdToken().getTokenValue());
System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getAudienceAsList());
System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());
```

in that way, the `AccessToken` and `IdToken` diverge and thats why i set them up different.




FWIW, he only reason i extended `AccessToken` is so that i can plumb it inot a `Credential` object for use like this

```java
// Invoke the API
GenericUrl genericUrl = new GenericUrl("https://myapp-6w42z6vi3q-uc.a.run.app");
HttpCredentialsAdapter adapter = new HttpCredentialsAdapter(tokenCredential);
HttpTransport transport = new NetHttpTransport();
HttpRequest request = transport.createRequestFactory(adapter).buildGetRequest(genericUrl);
request.setThrowExceptionOnExecuteError(false);
HttpResponse response = request.execute();
```

>> note, an `IdToken` cannot currently be used against a GCP service like GCS, Pub/Sub  (a self-issued `JWTAccessToken` can be, however...(some conjecture on my part: i suspect someday a googleissued ID token can be used in a similar way to `JWTAccessToken` to interact with a googleAPI...)

* I've removed `IdTokenProvider.IdTokenProviderException`. LMK if it shoudl remain (i bubble everything via `IOException`)

@broady

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +158,49 @@ public AccessToken refreshAccessToken() throws IOException {
+      if (options.contains(ID_TOKEN_LICENSES_TRUE))
+        optionalParams = optionalParams + "&licenses=TRUE";
+    }
+    String documentURL = getIdentityDocumentUrl() + "?audience=" + targetAudience + optionalParams;
```
`GenericUrl` should be used to build up this url instead of concatenating strings.

```java
GenericUrl documentUrl = new GenericUrl(getIdentityDocumentUrl());
documentUrl.set("audience", targetAudience);
// set other optional params
```

salrashid123 replied:
added in as url

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -138,7 +142,7 @@ public AccessToken refreshAccessToken() throws IOException {
     InputStream content = response.getContent();
     if (content == null) {
-      // Throw explicitly here on empty content to avoid NullPointerException from parseAs call.
+     // Throw explicitly here on empty content to avoid NullPointerException from parseAs call.
```
The indentation here was correct before.

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2015, Google Inc. All rights reserved.
```
```suggestion
 * Copyright 2019, Google LLC
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,89 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
```
```suggestion
 *    * Neither the name of Google LLC nor the names of its
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright 2018, Google Inc. All rights reserved.
```
```suggestion
 * Copyright 2019, Google LLC
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
```
```suggestion
 *    * Neither the name of Google LLC nor the names of its
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
```
```suggestion
 * Copyright 2019, Google LLC
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,114 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
```
```suggestion
 *    * Neither the name of Google LLC nor the names of its
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,114 @@
+    @Override
+    public int hashCode() {
+      return Objects.hash(getMessage(), getCause());
+    }
```
Do we actually need to override any of this?

salrashid123 replied:
no, removed that bit and unused `IdTokenException`

chingor13 commented:
> FWIW, he only reason i extended AccessToken is so that i can plumb it into a Credential object for use like this

The `IdTokenCredential` can be used in the `HttpCredentialsAdapter` and should be setting the "Authorization: Bearer [tokenValue]" header if installed as a `HttpRequestInitializer`.

chingor13 commented:
As for whether `IdToken` should extend `AccessToken`, OpenID Connect extends OAuth 2.0 so an `IdToken` is probably find extending `AccessToken`.

Perhaps `IdTokenCredentials` should extend `OAuth2Credentials` instead of `GoogleCredentials` given that it's not currently able to be used against Google services and you cannot call `createScoped` or `createDelegated`.

salrashid123 commented:
ok, i changed to

```java
public class IdTokenCredentials extends OAuth2Credentials {
```

and verified the credentials are usable in the same way.  I also see the raw token accessible in a couple ways:

```java
System.out.println(tokenCredential.getAccessToken().getTokenValue());               
System.out.println(tokenCredential.getIdToken().getTokenValue());
System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getAudienceAsList());
System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -370,7 +372,44 @@ public boolean isRequired(HttpResponse response) {
+   * @param options        List of Credential specific options for for the
+   *                       token. Currently unused for ServiceAccountCredentials.
+   * @throws IOException   if the attempt to get an IdToken failed
+   * @return IdToken object which includes the raw id_token, expiration and
```
Can we add a `@see` link for the docs for this API

salrashid123 replied:
we really don't have any documentation that covers this flow..I am working on publishing them (internal ref b/136495350)

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -125,4 +128,73 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
+
+  static IdToken getIdToken(String serviceAccountEmail, Credentials credentials, HttpTransport transport,
+      String targetAudience, boolean includeEmail, Map<String, ?> additionalFields) throws IOException {
+    IdToken token;
```
This whole method can be combined with getOIDToken

salrashid123 replied:
done

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -125,4 +128,73 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
+   * @param targetAudience      the audience the issued ID token should include
+   * @param additionalFields    additional fields to send in the IAM call
+   * @return New IdToken issed to the serviceAccount.
+   * @throws IOException if the IdToken cannot be issued.
```
Can we add a `@see` link to the developer docs for this API

salrashid123 replied:
done

salrashid123 commented on oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java:
```diff
@@ -256,6 +256,34 @@ public AccessToken refreshAccessToken() throws IOException {
+  @Override
+  public IdToken idTokenWithAudience(String targetAudience, List<IdTokenProvider.Option> options) throws IOException {
+    boolean includeEmail = false;
+    if (options != null)
```
ok

salrashid123 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+  }
+
+  public IdToken getIdToken() {
+    return this.idToken;
```
done

salrashid123 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,89 @@
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof AccessToken)) {
```
done

salrashid123 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    if (!(this.sourceCredentials instanceof IdTokenProvider)) {
```
yeah, i tested it by in a code snippet where i removed thee explicit `sourcecredential.refresh()` bit and just relied on the underlying object.

then ran the code, waited over just hr for the underlying expiration and then saw the new token get generated

```java
               System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());
Thread.sleep(61 * 60 * 1000);
// after this, the credential got refreshed              
response = request.execute();
r = response.parseAsString();
System.out.println(r);               
System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());

```

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+  /**
+   * Returns an Google Id Token from the metadata server on ComputeEngine.
+   * 
+   * @param targetAudience The aud: field the IdToken should include.
```
nit: per Javadoc guidelines from Oracle, no initial caps (The --> the) and no period at end unless there's more than one sentence

salrashid123 replied:
ok, removing inital capitals

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+   * Returns an Google Id Token from the metadata server on ComputeEngine.
+   * 
+   * @param targetAudience The aud: field the IdToken should include.
+   * @param options        List of Credential specific options for for the
```
We don't indent like this since it's too hard for autoformatters to maintain

salrashid123 replied:
ok, removed indent (vs code's java 'autoformatter' does this type of indents for you)

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+   *                       If no options are set, the default
+   *                       are "&amp;format=standard&amp;licenses=false"
+   * @throws IOException   if the attempt to get an IdToken failed
+   * @return IdToken object which includes the raw id_token, JsonWebSignature.
```
no period

salrashid123 replied:
ok

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+  public IdToken idTokenWithAudience(String targetAudience, List<IdTokenProvider.Option> options) throws IOException {
+    GenericUrl documentURL = new GenericUrl(getIdentityDocumentUrl());
+    if (options != null) {
+      if (options.contains(IdTokenProvider.Option.FORMAT_FULL))
```
google style always uses brace on if statements, even single line ones

salrashid123 replied:
ok

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+   */
+  @Override
+  public IdToken idTokenWithAudience(String targetAudience, List<IdTokenProvider.Option> options) throws IOException {
+    GenericUrl documentURL = new GenericUrl(getIdentityDocumentUrl());
```
documentUrl per Google style https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names

salrashid123 replied:
ok

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+        documentURL.set("license","TRUE");
+    }
+    documentURL.set("audience", targetAudience);
+    HttpResponse response = null;
```
combine this and next line

elharo replied:
Ping

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,44 @@ public AccessToken refreshAccessToken() throws IOException {
+      throw new IOException("Empty content from metadata token server request.");
+    }
+    String rawToken = response.parseAsString();
+    JsonWebSignature jws =  JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, rawToken);
```
no abbreviated method name please, per Google style. `signature` is fine

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -229,6 +271,11 @@ public static String getServiceAccountsUrl() {
+
   @Override
   public int hashCode() {
     return Objects.hash(transportFactoryClassName);
```
not for this PR, but this looks really wonky. Note to myself: can this possibly be correct?

elharo commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -109,11 +112,11 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
       String errorMessage = OAuth2Utils.validateString(error, "message", PARSE_ERROR_MESSAGE);
       throw new IOException(String.format("Error code %s trying to sign provided bytes: %s",
-              statusCode, errorMessage));
+                statusCode, errorMessage));
```
continuation lines are indented 4 spaces, not 8. This is one of the areas where Google style deviates from Oracle style.

salrashid123 replied:
ok, i think i got the indents here right; PTAL

elharo replied:
doesn't look right. Are you using tabs?

elharo commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -125,4 +128,67 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
+    for (Map.Entry<String, ?> entry : additionalFields.entrySet()) {
+      idTokenRequest.set(entry.getKey(), entry.getValue());
+    }
+    JsonHttpContent signContent = new JsonHttpContent(OAuth2Utils.JSON_FACTORY, idTokenRequest);
```
signedContent?

salrashid123 replied:
ok

elharo commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -125,4 +128,67 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
+   * Returns an IdToken issued to the serviceAccount with a specified
+   * targetAudience.
+   *
+   * @param serviceAccountEmail the email address for the service account to get
```
probably ID token if it's not code since Id and ID are different things

salrashid123 replied:
ok

elharo commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -125,4 +128,67 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
+      String errorMessage = OAuth2Utils.validateString(error, "message", PARSE_ERROR_MESSAGE);
+      throw new IOException(String.format("Error code %s trying to getIDToken: %s", statusCode, errorMessage));
+    }
+    if (statusCode != HttpStatusCodes.STATUS_CODE_OK) {
```
should this be any 200 level response?

salrashid123 replied:
i'm not absolutely sure but i've never seen responses range 2xx  from this api layer:
 ref: https://cloud.google.com/apis/design/errors#handling_errors
 but storage mentions a bit of that here: https://cloud.google.com/storage/docs/json_api/v1/status-codes

(the google apis endpoint iamcredentials uses is older one (similar to drive api, calendar api, etc)

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -64,7 +68,7 @@
  * </p>
  */
-public class ComputeEngineCredentials extends GoogleCredentials implements ServiceAccountSigner {
+public class ComputeEngineCredentials extends GoogleCredentials implements ServiceAccountSigner, IdTokenProvider {
```
Is this going to require a major version bump? That is, will it break existing subclasses that do not implement IdTokenProvider?

salrashid123 replied:
@chingor13  

thye'll just inherit all these changes

chingor13 replied:
This shouldn't be a breaking change.

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,46 @@ public AccessToken refreshAccessToken() throws IOException {
   }
 
+  /**
+   * Returns an Google ID Token from the metadata server on ComputeEngine
```
nit: a Google ID

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,46 @@ public AccessToken refreshAccessToken() throws IOException {
+   * 
+   * @param targetAudience the aud: field the IdToken should include
+   * @param options list of Credential specific options for for the
+   * token. For example, an IDToken for a ComputeEngineCredential could have 
```
continuation lines should be indented four spaces past the asterisk. google-java-format plugins for Eclipse and IntelliJ should do this automatically

salrashid123 replied:
i'm using vscode. is i dind't find a plugin formatter for that.
the spaces in this example start off at the text possiton for the description of the param
https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java#L222

if the guide is just 4 spaces post `*`, thats fine, i'll alter it but i'm looking at the existing sample as reference

chingor13 replied:
We enabled a maven plugin to assist with formatting. Please merge from master and run `mvn com.coveo:fmt-maven-plugin:format`

salrashid123 replied:
ok, i ran that and added in some testcases for all the credential types.

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,46 @@ public AccessToken refreshAccessToken() throws IOException {
+   * is provided as a list option.  Valid option values are:
+   * IdTokenProvider.Option.FORMAT_FULL<br>
+   * IdTokenProvider.Option.LICENSES_TRUE<br>
+   * If no options are set, the default
```
defaults are or default is

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -151,6 +155,46 @@ public AccessToken refreshAccessToken() throws IOException {
+        documentUrl.set("format", "full");
+      }
+      if (options.contains(IdTokenProvider.Option.LICENSES_TRUE)) {
+        // if license will only get returned if format is also full
```
delete "if"

salrashid123 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,277 @@
+  private static final long serialVersionUID = -2133257318957588431L;
+  private static final String CLOUD_PLATFORM_SCOPE = "https://www.googleapis.com/auth/cloud-platform";
+
+  private GoogleCredentials sourceCredentials;
```
i'm marking this sequence as resolved as this squence is ultimately if 

```
public class IdToken extends AccessToken
```  
is done as described in https://github.com/googleapis/google-auth-library-java/pull/303#issuecomment-515244895

chingor13 replied:
Not all `GoogleCredentials` will work for refreshing the `IdToken`. We are not using any methods from `GoogleCredentials` so this should be an `IdTokenProvider`.

salrashid123 replied:
Changing it will prevent a customer from using implict `GoogleCredential` which may be an underlying supported one.

that is, right now i can do something like this
```
export GOOGLE_APPLICATION_CREDENTIALS=/path/to/svc.json
```

then

```java
GoogleCredentials adcCreds = GoogleCredentials.getApplicationDefault();
IdTokenCredentials tokenCredential = IdTokenCredentials.create(adcCreds, targetAudience);
```

works

but if i use a userCredentials (which doens't implement it)

the follwing is thrown
```
IOException java.io.IOException: Provided sourceToken does not implement IdTokenProvider
```
--

I can do either but see its pretty convenient to allow both `GoogleCredentials` and explicit ones like `ComputeCredentials`, etc

chingor13 replied:
Since `IdTokenCredentials` is not a `GoogleCredentials` subclass, it doesn't really make a whole lot of sense to try to accept all `GoogleCredentials` types.

current:
```java
GoogleCredentials sourceCredentials = GoogleCredentials.getApplicationDefault();
IdTokenCredentials credentials = IdTokenCredentials.create(sourceCredentials, targetAudience);

// use the credentials
try {
  // fails when we try to use the credentials
  credentials.getRequestMetadata();
} catch (IOException ex) {
  if (ex.getMessage().equals("Provided sourceToken does not implement IdTokenProvider") {
    // wrong credentials type but nothing you can do about it
  } else {
    // handle other IOExceptions
  }
}
```

If we do allow `IdTokenCredentials.create(GoogleCredentials, String)`, we would want it to fail early with something like an `IllegalArgumentException` rather than waiting until it's used. 

with IllegalArgument:
```java
GoogleCredentials sourceCredentials = GoogleCredentials.getApplicationDefault();
IdTokenCredentials credentials;
try {
  credentials = IdTokenCredentials.create(sourceCredentials, targetAudience);
} catch (IllegalArgumentException e) {
  // test that IllegalArgumentException is because of invalid type and not some other possible illegal argument
}

// use the credentials
try {
  credentials.getRequestMetadata();
} catch (IOException ex) {
  // handle a request error or credentials error
}
```

I'm not a fan of this approach because the error handling is difficult. You'd rather have the user explicitly have to handle an unexpected credentials type if they are trying to use an IdTokenCredentials with an incompatible type.

change to IdTokenProvider:
```java
GoogleCredentials sourceCredentials = GoogleCredentials.getApplicationDefault();
if (!sourceCredentials instanceof IdTokenProvider) {
  // handle error message
}
IdTokenCredentials credentials = IdTokenCredentials.create((IdTokenProvider) sourceCredentials, targetAudience);

// use the credentials
try {
  credentials.getRequestMetadata();
} catch (IOException ex) {
  // handle a request error or credentials error
}
```

salrashid123 replied:
fair enough; i'll keep `IdTokenProvider`

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,88 @@
+   * @param tokenValue     String representation of the Id token.
+   * @param jws            JsonWebSignature as object
+   */
+  public IdToken(String tokenValue, JsonWebSignature jws) {
```
Everywhere we instantiate an `IdToken`, we are immediately preceding that line with `JsonWebSignature.parse(tokenValue)`. That logic should be handled in this constructor or via a static `IdToken.create(String)` method.

salrashid123 replied:
yeah...thats what i initially wanted to do but ran into a bit of an issue:

since i'm extending `IdToken` as 
```
public class IdToken extends AccessToken
```

i need to `super()` in the constructor for `IdToken` which means i need to pass in

```java
public AccessToken(String tokenValue, Date expirationTime)
```

that is, i need to pass in the raw token value and a `Date` object defining the expiration time.

The only way i can get the expiration time out of the jwt is to parse the JWT first.

There maybe an easier way to work around this properly but this bit of awkwardness is due to  using AccessToken as a basis

chingor13 replied:
Can we do something like:
```java
public IdToken(String tokenValue) {
  try {
    this.jws = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, tokenValue);
  } catch (IOException e) {
    throw new IllegalArgumentException("Invalid JsonWebSignature token provided", e);
  }
  super(tokenValue, new Date(this.jws.getPayload().getExpirationTimeSeconds() * 1000));
}
```

or:
```java
private IdToken(String tokenValue, JsonWebSignature jws) {
  // same thing you have
}

public static IdToken create(String tokenValue) throws IOException {
  return create(tokenValue, OAuth2Utils.JSON_FACTORY);
}

public static IdToken create(String tokenValue, JsonFactory jsonFactory) throws IOException {
  return new IdToken(tokenValue, JsonWebSignature.parse(jsonFactory, tokenValue);
}
```

salrashid123 replied:
ok, 

```java
  public IdToken(String tokenValue, JsonWebSignature jws) {
    super(tokenValue, new Date(jws.getPayload().getExpirationTimeSeconds() * 1000));
    this.jws = jws;
  }

  public static IdToken create(String tokenValue) throws IOException {
    return create(tokenValue, OAuth2Utils.JSON_FACTORY);
  }
  
  public static IdToken create(String tokenValue, JsonFactory jsonFactory) throws IOException {
    return new IdToken(tokenValue, JsonWebSignature.parse(jsonFactory, tokenValue));
  }  
```

works; i think in the ifrst one had the same issue with `super()` not getting called first

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,276 @@
+  private String targetAudience;
+  private List<IdTokenProvider.Option> options;
+
+  private transient HttpTransportFactory transportFactory;
```
This looks unused - we are delegating the fetching to the source `GoogleCredentials` (which should be `IdTokenProvider` instead). We can remove this and the `transportFactoryClassName`.

salrashid123 replied:
removed `transportFactoryClassName`  but see previous comment on using `IdTokenProvider` instead of `GoogleCredentials`

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,88 @@
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(super.getTokenValue(), jws);
```
nit: In theory the raw token and the parsed JsonWebSignature should be referencing the same data.

salrashid123 replied:
yes, it should be but since i'm passing them in separately they maybe off.  i tried putting this check into the constructor just now

```java
  public IdToken(String tokenValue, JsonWebSignature jws) throws IOException{
    super(tokenValue, new Date(jws.getPayload().getExpirationTimeSeconds() * 1000));
    if (!tokenValue.equals(jws.toString())) {
      throw new IOException("Provided IdToken value does not match provided JsonWebSignature");
    }
    this.jws = jws;
  }
```

but `JsonWebSignature.toString()` returns a formatted json string and it doesn't expose an automatic way to return the encoded JWT (hint; it'd be a good api to have in `JsonWebSignature`...

elharo replied:
It would be easier to make this change in google3's monorepo. Nonetheless, do not rule out adding such a method to JsonWebSignature in google-http-java-client first and then using it here.

salrashid123 replied:
ok, I'll leave it as-is; we can always ad din that check for coverage later on

salrashid123 replied:
ref:
  https://github.com/googleapis/google-http-java-client/pull/757

elharo commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -157,6 +160,46 @@ public AccessToken refreshAccessToken() throws IOException {
+   *
+   * @param targetAudience the aud: field the IdToken should include
+   * @param options list of Credential specific options for for the token. For example, an IDToken
+   *     for a ComputeEngineCredential could have the full formated claims returned if
```
nit: formatted

elharo commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -137,4 +141,75 @@ private static String getSignature(
+
+    GenericJson responseData = response.parseAs(GenericJson.class);
+    String rawToken = OAuth2Utils.validateString(responseData, "token", PARSE_ERROR_MESSAGE);
+    JsonWebSignature jws = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, rawToken);
```
jws--> signature (no abbreviations per Google style)

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,88 @@
+
+  private static final long serialVersionUID = -8514239465808977353L;
+
+  private final JsonWebSignature jws;
```
jws --> signature

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  /**
+   * Returns the a Google OpenID Token with the provided audience field.
```
the a --> a

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,76 @@
+   * Returns the a Google OpenID Token with the provided audience field.
+   *
+   * @param targetAudience List of audiences the issued ID Token should be valid for. targetAudience
+   *     accepts a single string value (multiple audience are not supported)
```
audience --> audiences

elharo commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -437,6 +439,43 @@ public boolean isRequired(HttpResponse response) {
+    HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
+    HttpRequest request = requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), content);
+    request.setParser(new JsonObjectParser(jsonFactory));
+    HttpResponse response;
```
combine this with next line

elharo replied:
ping

elharo commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -437,6 +439,43 @@ public boolean isRequired(HttpResponse response) {
+    GenericData responseData = response.parseAs(GenericData.class);
+    String rawToken = OAuth2Utils.validateString(responseData, "id_token", PARSE_ERROR_PREFIX);
+
+    JsonWebSignature jws = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, rawToken);
```
signature

salrashid123 replied:
removed the section already from here

elharo commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -595,6 +634,38 @@ String createAssertion(JsonFactory jsonFactory, long currentTime, String audienc
+
+    payload.set("target_audience", targetAudience);
+
+    String assertion;
```
push this declaration and the return statement into the try block

elharo commented on oauth2_http/javatests/com/google/auth/oauth2/MockMetadataServerTransport.java:
```diff
@@ -145,6 +155,61 @@ public LowLevelHttpResponse execute() throws IOException {
+      }
+
+      // https://cloud.google.com/compute/docs/instances/verifying-instance-identity#token_format
+      URL u = new URL(url);
```
u --> url

salrashid123 replied:
i'm already using `url` as a variable so i collapsed it as
```
String query = (new URL(url)).getQuery();
```

elharo commented on oauth2_http/javatests/com/google/auth/oauth2/MockMetadataServerTransport.java:
```diff
@@ -145,6 +155,61 @@ public LowLevelHttpResponse execute() throws IOException {
+
+      // https://cloud.google.com/compute/docs/instances/verifying-instance-identity#token_format
+      URL u = new URL(url);
+      Map<String, String> query_pairs = new HashMap<String, String>();
```
queryPairs per Google style

salrashid123 replied:
done...

side note: i found it suprising parsing queryparameters was so manual....

elharo commented on oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java:
```diff
@@ -517,6 +583,60 @@ public void refreshAccessToken_failsNotFoundError() throws IOException {
 
+  @Test
+  public void idTokenWithAudience_correct() throws IOException {
+    final String accessToken1 = "1/MkSJoj1xsli0AccessToken_NKPY2";
```
google style is to not uise final on local variables unless it's required

salrashid123 replied:
yah, its all over the place in the existing test case;  i only removed it in the section i added.

elharo commented on oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java:
```diff
@@ -517,6 +583,60 @@ public void refreshAccessToken_failsNotFoundError() throws IOException {
+
+  @Test
+  public void idTokenWithAudience_incorrect() throws IOException {
+    final String accessToken1 = "1/MkSJoj1xsli0AccessToken_NKPY2";
```
google style is to not uise final on local variables unless it's required

salrashid123 replied:
pls see above comment

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,225 @@
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    return ((IdTokenProvider) this.sourceCredentials).idTokenWithAudience(targetAudience, options);
```
This cast is no longer necessary.

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,225 @@
+  private static final String CLOUD_PLATFORM_SCOPE =
+      "https://www.googleapis.com/auth/cloud-platform";
+
+  private IdTokenProvider sourceCredentials;
```
We should rename this and the builder methods to `idTokenProvider`

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java:
```diff
@@ -157,6 +160,46 @@ public AccessToken refreshAccessToken() throws IOException {
+    }
+    String rawToken = response.parseAsString();
+    JsonWebSignature jws = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, rawToken);
+    return new IdToken(rawToken, jws);
```
No longer need to do the parsing here.

```suggestion
    return IdToken.create(rawToken);
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IamUtils.java:
```diff
@@ -137,4 +141,75 @@ private static String getSignature(
+    GenericJson responseData = response.parseAs(GenericJson.class);
+    String rawToken = OAuth2Utils.validateString(responseData, "token", PARSE_ERROR_MESSAGE);
+    JsonWebSignature signature = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, rawToken);
+    return new IdToken(rawToken, signature);
```
No longer need to do the parsing here

```suggestion
    return IdToken.create(rawToken);
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -437,6 +439,43 @@ public boolean isRequired(HttpResponse response) {
+    String rawToken = OAuth2Utils.validateString(responseData, "id_token", PARSE_ERROR_PREFIX);
+
+    JsonWebSignature jws = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, rawToken);
+    return new IdToken(rawToken, jws);
```
No longer need to do the parsing here

```suggestion
    return IdToken.create(rawToken);
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,98 @@
+   * @param tokenValue String representation of the Id token.
+   * @param jsonWebSignature JsonWebSignature as object
+   */
+  public IdToken(String tokenValue, JsonWebSignature jsonWebSignature) {
```
Consider making this private as the `tokenValue` and `jsonWebSignature` should be the same content. It would be unexpected behavior, but possible, for a user to set incompatible values.

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,98 @@
+import com.google.common.base.MoreObjects;
+
+/**
+ * Represents a temporary IdToken and its JSONWebSingature object
```
JsonWebSignature

salrashid123 commented:
ok, added  in `@Beta` annotations to `idTokenWithAudience` as well as the new classes added into this PR

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,229 @@
+  }
+
+  public Builder toBuilder() {
+    return new Builder();
```
This should return a builder with the current options set.

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,229 @@
+   *     tokens.
+   * @return IdTokenCredential
+   */
+  public static IdTokenCredentials create(
```
I don't think we should provide this (or the version without options) when the builder is available. It's not adding any convenience, is extra surface area, and could be confusing about which one to use.

If we are confident the interface will never be expanded, we could prefer the static create over the builder, but the builder is more flexible for the future.

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,100 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google LLC. nor the names of its
```
```suggestion
 *    * Neither the name of Google LLC nor the names of its
```

salrashid123 commented:
@chingor13 

 I think the last merge overwrote some of my changes somehow..
https://github.com/googleapis/google-auth-library-java/pull/303/commits/4445ee7d1c52c88db5ddb497759c4b1018088d68

chingor13 commented:
Ah `IAM_ID_TOKEN_ENDPOINT` in the mock transport was unused in master so it was removed in another branch.

salrashid123 commented:
Is there anything i should do to add this in or can you remerge?

@saturnism  
@grayside

chingor13 commented:
Please put the constant back in this branch. It was not a merge conflict.

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,199 @@
+    }
+    IdTokenCredentials other = (IdTokenCredentials) obj;
+    return Objects.equals(this.idTokenProvider, other.idTokenProvider)
+        && Objects.equals(this.targetAudience, other.targetAudience);
```
Bug: equals depends on targetAudience but hashCode does not

salrashid123 replied:
ok, so added in like this

```java
  @Override
  public int hashCode() {
    return Objects.hash(idTokenProvider, targetAudience);
  }
```

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java:
```diff
@@ -0,0 +1,199 @@
+
+  private static final long serialVersionUID = -2133257318957588431L;
+
+  private IdTokenProvider idTokenProvider;
```
All of these are nullable. Is that what you want?

salrashid123 replied:
so do the checks here?
```java
  private IdTokenCredentials(Builder builder) {
    this.idTokenProvider = Preconditions.checkNotNull(builder.getIdTokenProvider());
    this.targetAudience = Preconditions.checkNotNull(builder.getTargetAudience());
    this.options = Preconditions.checkNotNull(builder.getOptions());
  }
```

elharo replied:
That works.

salrashid123 replied:
ok, the options can be nullable so removed the condition checks on that

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.List;
+
+/** Interface for an Google OIDC token Provider. This type represents a google issued OIDC token. */
+public interface IdTokenProvider {
```
`@Beta`

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,76 @@
+import java.io.IOException;
+import java.util.List;
+
+/** Interface for an Google OIDC token Provider. This type represents a google issued OIDC token. */
```
provider (lower case)

elharo commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,76 @@
+
+    private String option;
+
+    Option(String option) {
```
private?

salrashid123 replied:
not sure i follow, it should be  private yeah?  its just the internal artificat (or should i just rename that field though)

elharo replied:
Yes, unless there's something I'm missing the constructor could be private.

salrashid123 replied:
oh, ok, i misinterpreted the comment to be for the variable (not constructor).  marked private

elharo commented on oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java:
```diff
@@ -275,6 +277,33 @@ public AccessToken refreshAccessToken() throws IOException {
+  /**
+   * Returns an IdToken for the current Credential
+   *
+   * @param targetAudience The audience field for the issued ID Token
```
nit: The --> the

elharo commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -438,6 +440,43 @@ public boolean isRequired(HttpResponse response) {
   }
 
+  /**
+   * Returns a Google Id Token from the metadata server on ComputeEngine.
```
Id --> ID

elharo commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -615,6 +654,38 @@ String createAssertion(JsonFactory jsonFactory, long currentTime, String audienc
+    try {
+      payload.set("target_audience", targetAudience);
+
+      String assertion;      
```
combine with next line

elharo commented on oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java:
```diff
@@ -275,6 +277,33 @@ public AccessToken refreshAccessToken() throws IOException {
   }
 
+  /**
+   * Returns an IdToken for the current Credential
```
period

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdTokenProvider.java:
```diff
@@ -0,0 +1,78 @@
+
+  /**
+   * Enum of various credential-specific options to apply to the token.
+   * <li><b>ComputeEngineCredentials</b>:
```
`<li>` needs to be enclosed in `<ul></ul>`. They should also be closed.

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/ImpersonatedCredentialsTest.java:
```diff
@@ -85,6 +86,21 @@
+          + "iaXNzIjoiaHR0cHM6Ly9hY2NvdW50cy5nb29nbGUuY29tIiwic3ViIjoiMTAyMTAxNTUwODM0MjAwNzA4NTY4In"
+          + "0.redacted";
+
+  public static final String tokenWithEmail =
```
Constants should be uppercased and underscored.

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/ImpersonatedCredentialsTest.java:
```diff
@@ -85,6 +86,21 @@
           + "ADj3e1YhMVdjJW5jqwlD/VNddGjgzyunmiZg0uOXsHXbytYmsA545S8KRQFaJKFXYYFo2kOjqOiC1T2cAzMDjCQ"
           + "==\n-----END PRIVATE KEY-----\n";
 
+  public static final String standardIdToken =
```
Constants should be uppercased and underscored.

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/IdTokenCredentialsTest.java:
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2015, Google Inc. All rights reserved.
```
```suggestion
 * Copyright 2019 Google LLC
```

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:
```diff
@@ -615,6 +653,38 @@ String createAssertion(JsonFactory jsonFactory, long currentTime, String audienc
     return assertion;
   }
 
+  String createAssertionForIdToken(
```
Please add `@VisibleForTesting` annotation here

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,113 @@
+    this.jsonWebSignature = jsonWebSignature;
+  }
+
+  public static IdToken create(String tokenValue) throws IOException {
```
Please add javadoc here

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,113 @@
+    return create(tokenValue, OAuth2Utils.JSON_FACTORY);
+  }
+
+  public static IdToken create(String tokenValue, JsonFactory jsonFactory) throws IOException {
```
Please add javadoc here

chingor13 commented on oauth2_http/java/com/google/auth/oauth2/IdToken.java:
```diff
@@ -0,0 +1,113 @@
+        && Objects.equals(this.jsonWebSignature.getPayload(), other.jsonWebSignature.getPayload());
+  }
+
+  private void writeObject(ObjectOutputStream oos) throws IOException {
```
Do we need `oos.defaultWriteObject()` here?

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/ComputeEngineCredentialsTest.java:
```diff
@@ -64,6 +68,38 @@
 
   private static final URI CALL_URI = URI.create("http://googleapis.com/testapi/v1/foo");
 
+  public static final String standardIdToken =
```
Constants should be uppercased and underscored.

chingor13 replied:
It might be worthwhile to add a comment on what's in these tokens so other's don't need to parse them to see the format.

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/ComputeEngineCredentialsTest.java:
```diff
@@ -64,6 +68,38 @@
+          + "aXNzIjoiaHR0cHM6Ly9hY2NvdW50cy5nb29nbGUuY29tIiwic3ViIjoiMTAyMTAxNTUwODM0MjAwNzA4NTY4In0"
+          + ".redacted";
+
+  public static final String fullIdToken =
```
Constants should be uppercased and underscored.

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/ComputeEngineCredentialsTest.java:
```diff
@@ -64,6 +68,38 @@
+          + "XMtY2VudHJhbDEtYSJ9fSwiaWF0IjoxNTY0NTE1ODk2LCJpc3MiOiJodHRwczovL2FjY291bnRzLmdvb2dsZS5jb"
+          + "20iLCJzdWIiOiIxMTIxNzkwNjI3MjAzOTEzMDU4ODUifQ.redacted";
+
+  public static final String fullIdTokenWithLicense =
```
Constants should be uppercased and underscored.

chingor13 commented on oauth2_http/javatests/com/google/auth/oauth2/IdTokenCredentialsTest.java:
```diff
@@ -0,0 +1,116 @@
+/*
+ * 2019 Google LLC
```
This still needs the copyright
```suggestion
 * Copyright 2019 Google LLC
```

salrashid123 replied:
done; could our formatter `mvn com.coveo:fmt-maven-plugin:format`  do all these checks?

chingor13 replied:
`maven-checkstyle-plugin` should verify the copyright headers if you run the tests.

salrashid123 commented:
ok,i reverified all the credential types work as intended;
- checked refresh works (ie., init one credential, use it, sleep for 1 hr then resuse the credential)

Note, the `UserCredentials` (ones based on an enduser three legged auth), does not implement the  required interface so the cast won't work 

```java
GoogleCredentials adcCreds = GoogleCredentials.getApplicationDefault();              
IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
                         .setIdTokenProvider((IdTokenProvider) adcCreds)
                         .setTargetAudience(targetAudience).build();
```
will throw
```java
Caused by: java.lang.ClassCastException: class com.google.auth.oauth2.UserCredentials cannot be cast to class com.google.auth.oauth2.IdTokenProvider (com.google.auth.oauth2.UserCredentials and com.google.auth.oauth2.IdTokenProvider are in unnamed module of loader java.net.URLClassLoader @6345dc06)
```

(unless ofcourse the `GoogleCredential` is internally a `ServiceAccountCredential`.  Thid'd be working as intended though)

salrashid123 commented:
thanks; final note, you can use these id_tokens now directly inside gRPC channel credentials

https://github.com/salrashid123/grpc_google_id_tokens/blob/master/java/src/main/java/com/test/TestApp.java#L62

(i'll update my sample there once this change is live and shipped in a release)
